### PR TITLE
Discard the object body when doing metadata request to S3

### LINKED ISSUE #9932: Receiving S3 input stream abort warnings when using an S3 maven repository
### Expected Behavior
Executing a build with an S3 maven repository runs without warnings.

### Current Behavior
Executing a build with an S3 maven repository produces multiple instances of the following warning.

> WARNING: Not all bytes were read from the S3ObjectInputStream, aborting HTTP connection. This is likely an error and may result in sub-optimal behavior. Request only the bytes you need via a ranged GET or drain the input stream after use.

### Context

We are using S3 as a maven repository for some of our libraries. When executing builds that rely on these dependencies, the above warning is being output multiple times.

After some investigation, I believe that this is an issue with the getMetaData method on S3ResourceConnector in conjunction with the getMetaData method on S3Client. The getMetaData method of S3Client returns an S3Object instead of just the metadata. It appears this is because of an issue with the GetObjectMetadataRequest as seen in the comment shown below. The S3ResourceConnector then gets the metadata and finally closes the S3Object. This warning is occurring because S3Object's content is never read before it is closed.

S3Client
```java
public S3Object getMetaData(URI uri) {
    LOGGER.debug("Attempting to get s3 meta-data: [{}]", uri.toString());
    //Would typically use GetObjectMetadataRequest but it does not work with v4 signatures
    return doGetS3Object(uri, true);
}
```

S3ResourceConnector
```java
public ExternalResourceMetaData getMetaData(URI location, boolean revalidate) {
    LOGGER.debug("Attempting to get resource metadata: {}", location);
    S3Object s3Object = s3Client.getMetaData(location);
    if (s3Object == null) {
        return null;
    }
    try {
        ObjectMetadata objectMetadata = s3Object.getObjectMetadata();
        return new DefaultExternalResourceMetaData(location,
            objectMetadata.getLastModified().getTime(),
            objectMetadata.getContentLength(),
            objectMetadata.getContentType(),
            objectMetadata.getETag(),
            null); // Passing null for sha1 - TODO - consider using the etag which is an MD5 hash of the file (when less than 5Gb)
    } finally {
        IoActions.closeQuietly(s3Object);
    }
}
```

I was able to confirm this with a simple test. This replicates the calls as they are above. When executed this will produce the warning. When explicitly calling abort on the object content prior to calling close, it does appear to stop this warning from occurring. Alternatively, reading through all of the bytes of the object also clears this warning. 

```java
public static void main(String[] args) throws Exception {
    GetObjectRequest gor = new GetObjectRequest("bucket", "key");
    S3Object s3Object = AmazonS3Client.builder().build().getObject(gor);
    ObjectMetadata objectMetadata = s3Object.getObjectMetadata();
    objectMetadata.getLastModified().getTime();
    objectMetadata.getContentLength();
    objectMetadata.getContentType();
    objectMetadata.getETag();
//        s3Object.getObjectContent().abort();

    s3Object.close();
}
```

### PR DESCRIPTION
**Pull request by mlopatkin**
Otherwise the client library issues a warning about potential waste.
There is waste but only 1 byte (Range header is used). I've kept an
info message for the cases where more than 1 byte is downloaded - just
in case if someone needs to investigate issues with the S3-compatible
service that doesn't understand Range headers and returns the whole
object instead.

Fixes #9932

### CODE REVIEW

**Comment by mlopatkin** [comment]
@bot-gradle test this

File: subprojects/resources-s3/src/main/java/org/gradle/internal/resource/transport/aws/s3/S3ResourceConnector.java [hunk 1]
```diff
@@ -72,6 +73,22 @@ public ExternalResourceMetaData getMetaData(URI location, boolean revalidate) {
...
+        }
+    }
+
+    private static void discardEmptyContentAndClose(S3Object s3Object) {
+        // Consume the content stream to avoid warning from S3 SDK. The response should have only 1 byte there because Range header was specified.
+        try {
+            long downloadedContentLength = ByteStreams.exhaust(s3Object.getObjectContent());
+            if (downloadedContentLength > 1L) {
+                // This may happen if the endpoint ignores Range HTTP header for whatever reason.
+                LOGGER.info("Downloaded {} bytes of the object content for metadata request which is too much.", downloadedContentLength);
```
Comment (ljacomet) [inline subprojects/resources-s3/src/main/java/org/gradle/internal/resource/transport/aws/s3/S3ResourceConnector.java:87]: I would make this a `DEBUG` level message as well.

**Comment by mlopatkin** [comment]
@bot-gradle test this

**Comment by mlopatkin** [comment]
@bot-gradle test QF

**Comment by big-guy** [comment]
@bot-gradle test and merge

**Comment by big-guy** [comment]
@mlopatkin is there some way of triggering this behavior in a test?

**Pull request merged**